### DISCUSSION TITLE
"default" Attribute to give to interface implementations.

### DISCUSSION THREAD

**Discussion by MkazemAkhgary** [dotnet/csharplang#1104] [General]
I have a custom list like

     class MyList<T> : IEnumerable<IList<T>>,  IEnumerable<IReadOnlyList<T>>,  IEnumerable<Foo<T>>, ...

because multiple interfaces are implemented, I cant use linq easily. so I have to write something like

    mylist.Where<Foo<T>>(foo => true);

It would be nice if I could give priority to interface implementations which tells compiler, if you cant infer type from usage, pick this one.

     [DefaultInterface(typeof(IEnumerable<Foo<T>>))]
     class MyList<T> : IEnumerable<IList<T>>,  IEnumerable<IReadOnlyList<T>>,  IEnumerable<Foo<T>>

this probably is hard to implement, but I think its worth it to help inferring generic type arguments from usage.

By the way I can always cast my list into preferred ienumerable before writing query, but since its part of API, putting this limitation away would be nice.

**Comment by iam3yal** [comment]
<s>@MkazemAkhgary Related #278.</s>

**Comment by jnm2** [comment]
Just curious: why not make it `IEnumerable<MyCustomType<T>>` which implements IList and IReadOnlyList etc? I've never been in this situation without being able to find a more desirable design.

**Comment by ufcpp** [comment]
different but similar to https://github.com/dotnet/csharplang/issues/964

**Comment by Mafii** [comment]
@jnm2 I agree. The problem that op is trying to solve with the default attribute can be solved by proper design, I don't think this feature (DefaultInterface) would benefit a lot of people:

    class Foot<T> : IList<T>, IReadOnlyList<T>
    {
        ... 
    }

and then just:

    class MyList<T> : IEnumerable<Foo<T>> 
    {
      ...
    }

should work.

**Comment by MkazemAkhgary** [comment]
You are right, this is something that can be fixed with correct design.

**Comment by MkazemAkhgary** [comment]
I reopen this issue because of this scenario. 

```
class Foo : IFoo<T>
interface IFoo<T> : IReadOnlyFoo<T>, IEnumerable<IBar<T>>
interface IReadOnlyFoo<out T> : IEnumerable<IReadOnlyBar<T>>
```

how can I improve my design? 

if I deprecate enumerable from readonly interface it works, but I want that. (enumerating is basic function)
if I deprecate readonly interface all together it works, but I want covariance from it.
if I deprecate enumerable from IFoo<T> then I will only have enumerable on readonly Bars.
